# PanMAGs
This repository provides codes and files to reproduce data and figures from the manuscript "Pan-genomics of metagenome-assembled genomes suffers from fragmentation and incompleteness", by Tang Li and Yanbin Yin* (*corresponding author). Here, the Python and shell scripts cover downloading genome data, simulating metagenome-assembled genomes (MAGs) from complete genomes, analyzing pan-genome, performing Clusters of Orthologous Group (COG) functional annotations and comparing phylogenetic trees. The R codes cover reformatting data, generating plots and combining plots. <br> 
## Requirements
* [FastANI](https://github.com/ParBLiSS/FastANI)
* [Prokka](https://github.com/tseemann/prokka)
* [Blast+](https://github.com/ncbi/blast_plus_docs)
* [Roary](https://github.com/sanger-pathogens/Roary)
* [Anvi'o](https://github.com/merenlab/anvio)
* [BPGA](https://github.com/jbadavis/bpga)
## Data
The entire data generated in this study is too large to store on Github, some example data for _`Escherichia coli`_ are available in () for testing.
